src/api/api.ts

```diff
@@ -31,54 +81,244 @@ export class GhostApi {
       });
   }
 
+  private sleep(ms: number): Promise<void> {
+    return new Promise(resolve => setTimeout(resolve, ms));
+  }
+
+  /**
+   * @todo this doesn't work as for some reason the retried fetch times out
+   */
+  private async fetchRetryOnErrors<R extends GhostResponse>(
+    url: string | Request,
+    init: RequestInit = { method: 'GET' },
+    retryErrorTypes: string[],
+    verifyStatusWithConfigFetch: boolean = false
+  ): Promise<R> {
+    const makeAttempt = (retries: number): Promise<R> => {
+      debugLog(`Calling ${init.method} ${url}`);
+      return fetch(url, {
+        ...init,
+        ...{ timeout: getenv.int('GHOST_API_FETCH_TIMEOUT_MS', 10000) },
+      })
+        .then(res => res.json() as Promise<R>)
+        .then((res: R) => {
+          if (res.errors) {
+            if (res.errors.some(e => retryErrorTypes.includes(e.errorType))) {
+              const message = `${init.method} ${url} failed due to ${
+                res.errors.find(e => retryErrorTypes.includes(e.errorType))!
+                  .errorType
+              }`;
+
+              throw new RetryAdminRoute(message);
+            } else {
+              debugLog(res.errors);
+              throw new GhostAdminError(
+                `${
+                  init.method
+                } ${url} failed due to ${this.ghostErrorResponseToString(res)}`
+              );
+            }
+          }
+
+          return res;
+        })
+        .catch(async err => {
+          if (
+            err instanceof RetryAdminRoute ||
+            err.message.includes('network timeout')
+          ) {
+            debugLog(err.message);
+
+            if (retries > 5) {
+              throw new Error('Retry count exceeded');
+            }
+
+            debugLog(`Retrying ${init.method} ${url} after ${retries}s`);
+
+            await this.sleep(retries * 1000);
+
+            return makeAttempt(++retries);
+          }
+
+          throw err;
+        });
+    };
+
+    /**
+     * @hack note for some reason if we don't first test that Ghost is ready, if
+     * we attempt to make the call, and it fails, subsequent retries will always
+     * timeout. It appears to be some kind of error choke/throttle mechanism
+     * that is misbehaving.
+     */
+    if (verifyStatusWithConfigFetch) {
+      let attempts = 0;
+      let apiResponse;
+
+      do {
+        await this.sleep(attempts * 1000);
+        debugLog('Verifying ghost API readiness...');
+        apiResponse = await fetch(this.config.urls.configUrl);
+
+        if (apiResponse.status !== 200) {
+          const body: GhostResponse = await apiResponse.json();
+          debugLog(
+            `Ghost is not ready, got ${
+              apiResponse.status
+            }. Retrying in ${attempts}s.`,
+            this.ghostErrorResponseToString(body)
+          );
+        }
+        debugLog('Ghost API is ready.');
+        attempts++;
+      } while (apiResponse.status !== 200);
```

Comment:
Would be great to have a `MAX_ATTEMPTS` constant and use it with `attempts` in the while condition so we don't end up with infinite loop